python/cuspatial/cuspatial/core/binpreds/binpreds.py

```diff
@@ -0,0 +1,323 @@
+            df_result = (
+                result.groupby("idx").sum().sort_index()
+                == result.groupby("idx").count().sort_index()
+            )
+        elif contains_only_polygons(self.rhs) or contains_only_multipoints(
+            self.rhs
+        ):
+            partial_result = result.groupby("idx").sum()
+            df_result = (partial_result > 0) & (
+                partial_result < len(point_result)
+            )
+        else:
+            df_result = result.groupby("idx").sum() > 1
+        point_result = cudf.Series(
+            df_result["pip"], index=cudf.RangeIndex(0, len(df_result))
+        )
+        point_result.name = None
+        return point_result
+
+
+class IntersectsBinpred(ContainsProperlyBinpred):
+    def preprocess(self, lhs, rhs):
+        if contains_only_polygons(rhs):
+            (lhs, rhs) = (rhs, lhs)
+        return super().preprocess(lhs, rhs)
+
+
+class WithinBinpred(ContainsProperlyBinpred):
+    def preprocess(self, lhs, rhs):
+        if contains_only_polygons(rhs):
+            (lhs, rhs) = (rhs, lhs)
+        return super().preprocess(lhs, rhs)
+
+    def postprocess(self, point_indices, point_result):
+        """Postprocess the output GeoSeries to ensure that they are of the
+        correct type for the operation."""
+        result = cudf.DataFrame({"idx": point_indices, "pip": point_result})
+        df_result = result
+        # Discrete math recombination
+        if (
+            contains_only_linestrings(self.rhs)
+            or contains_only_polygons(self.rhs)
+            or contains_only_multipoints(self.rhs)
+        ):
+            # process for completed linestrings, polygons, and multipoints.
+            # Not necessary for points.
+            df_result = (
+                result.groupby("idx").sum().sort_index()
+                == result.groupby("idx").count().sort_index()
+            )
+        point_result = cudf.Series(
+            df_result["pip"], index=cudf.RangeIndex(0, len(df_result))
+        )
+        point_result.name = None
+        return point_result
+
+
+class EqualsBinpred(BinaryPredicate):
+    def preprocess(self, lhs, rhs):
+        return (lhs, rhs, None)
+
+    def postprocess(self, point_indices, point_result):
+        if isinstance(point_result, bool):
+            return point_result
+        elif len(point_result) == 1:
+            return point_result[0]
+
+    def _op(self, lhs, rhs):
+        """Compute the equals relationship between two GeoSeries."""
+        result = False
+        if contains_only_points(lhs):
+            result = lhs.points.xy.equals(rhs.points.xy)
+        elif contains_only_linestrings(lhs):
+            result = lhs.lines.xy.equals(rhs.lines.xy)
+        elif contains_only_polygons(lhs):
+            result = lhs.polygons.xy.equals(rhs.polygons.xy)
+        elif contains_only_multipoints(lhs):
+            result = lhs.multipoints.xy.equals(rhs.multipoints.xy)
```

Comment:
Is there a way to just get the geometry as the single type that the geoseries contains, rather than doing this multi-step if/else every time? 

```suggestion
        if contains_single_geometry_type(lhs):
            result = lhs.geometry.as_type(lhs.type).xy.equals(rhs.geometry.as_type(lhs.type).xy)
```

Reply:
We haven't implemented that yet. It is basically just remapping the `input_types` buffer of the `GeoColumn`.

Reply:
Seems like it would be very valuable.

Reply:
It will come in handy for fixing the bug in equals I identified that caused me to close this PR: I was using `.equals` instead of `.geom_equals`. :face_exhaling: